src/test/regress/regress_gp.c

```diff
@@ -2349,3 +2349,117 @@ gp_keepalives_check(PG_FUNCTION_ARGS) {
 
 	SRF_RETURN_DONE(funcctx);
 }
+
+typedef struct
+{
+	int cur_tuple_idx;
+	int cur_segment_idx;
+	int cur_segment_tuple_idx;
+	int n_segments;
+	int n_tuples;
+	struct pg_result **pg_results;
+} gp_mock_cdbdispatchcommand_status;
+
+/* 
+ * This test function mocks CdbDispatchCommand() with a customizable amount of
+ * tuples.
+ */
+PG_FUNCTION_INFO_V1(gp_mock_cdbdispatchcommand);
+Datum
+gp_mock_cdbdispatchcommand(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *func_ctx;
+	gp_mock_cdbdispatchcommand_status *my_status;
+
+	int32 arg_tuple_amount = PG_GETARG_INT32(0);
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+
+		func_ctx = SRF_FIRSTCALL_INIT();
+		oldcontext = MemoryContextSwitchTo(func_ctx->multi_call_memory_ctx);
+
+		my_status = (gp_mock_cdbdispatchcommand_status *) palloc0(
+			sizeof(gp_mock_cdbdispatchcommand_status));
+
+		if (Gp_role == GP_ROLE_DISPATCH)
+		{
+			int i;
+			CdbPgResults cdb_pgresults = {NULL, 0};
+
+			char *query =
+				psprintf("SELECT * FROM gp_mock_cdbdispatchcommand(%d)",
+						 arg_tuple_amount);
+
+			CdbDispatchCommand(query, DF_WITH_SNAPSHOT, &cdb_pgresults);
+
+			Assert(cdb_pgresults.numResults > 0);
+
+			for (i = 0; i < cdb_pgresults.numResults; i++)
+			{
+				Assert(PQresultStatus(cdb_pgresults.pg_results[i]) ==
+					   PGRES_TUPLES_OK);
+				my_status->n_tuples += PQntuples(cdb_pgresults.pg_results[i]);
+			}
+
+			my_status->n_segments = cdb_pgresults.numResults;
+			my_status->pg_results = cdb_pgresults.pg_results;
+
+			pfree(query);
+		}
+
+		func_ctx->user_fctx = (void *) my_status;
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	func_ctx = SRF_PERCALL_SETUP();
+	my_status = (gp_mock_cdbdispatchcommand_status *) func_ctx->user_fctx;
+
+	/* Generate fake tuples from every segment. */
+	while (my_status->cur_tuple_idx < arg_tuple_amount)
+	{
+		if (my_status->cur_tuple_idx++ == arg_tuple_amount)
+		{
+			my_status->cur_tuple_idx--;
+			break;
+		}
```

Comment:
Why so complicated? It's a simple for-cycle
```suggestion
	for (my_status->cur_tuple_idx = 0; my_status->cur_tuple_idx < arg_tuple_amount; my_status->cur_tuple_idx++)
	{
```